### ISSUE TITLE
Is [Singleton] attribute supported?

### ISSUE THREAD

**Issue by dmytro-gokun** [Azure/azure-functions-dotnet-worker#274]
Now, when version 1.0.0 is out, i've started trying to port some our functions. Stumbled upon this attribute on some of the functions. Is it expected to work?

**Comment by brettsam** [comment]
This will not work today as it only works with the in-proc model. We can use this issue to track the feasibility of adding this in the future.

**Comment by dmytro-gokun** [comment]
That would be really nice. I guess i'm not only one who would use this feature.

**Comment by jsheetzati** [comment]
Without this attribute, will a TimerTrigger that is ran under a multi-instance function app still only kick off once? We don't want our timers firing twice (we run our function app with 2 instances in prod)

https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-timer?tabs=csharp#scale-out

**Comment by brettsam** [comment]
Good question, @jsheetzati. 

The TimerTrigger is still a Singleton in multi-instance scenarios. This is because the TimerTrigger created by the Functions host is still a Singleton. The worker versions of the attributes (those that start with `Microsoft.Azure.Functions...`) are used to generate the metadata files that the host then uses to create the actual triggers. This is similar to how other out-of-proc languages work.

The lack of `[Singleton]` means that you cannot declare that one of your (for example) `[QueueTrigger]` functions is a singleton. But the triggers that already declare themselves as Singleton remain Singleton without any changes.

Just as a side note -- to enable `[Singleton]` for .NET out-of-proc would likely mean coming up with the correct metadata definitions so that all out-of-proc languages can support this. This work may be moved into the Functions Host repo when we add it to a Sprint.

**Comment by dmytro-gokun** [comment]
@brettsam What if we run a TimeTrigger function manually? Is "Singleton" behavior guaranteed in that case?

**Comment by dmytro-gokun** [comment]
@brettsam Yes, I mean "Test/Run". My understanding is that:
1) [Singleton] will prevent this manual task from running in parallel with some scheduled task that might be already running at the moment (and vice versa). 
2) [TimeTrigger] will only work as [Singleton] for scheduled runs, but not for manual ones.
Am I correct?

**Comment by BenjaBobs** [comment]
@brettsam Any news on this?
My team made use of the singleton attribute, and it not working in out-of-proc mode is what is holding us back from upgrading.

**Comment by SeanFeldman** [comment]
@brettsam, could you share some updates, especially with more people running into the same issue (#735, #1756)?

**Comment by essmd** [comment]
Any plan or roadmap on this imho essential feature? I did not find anything like a detailed overview of the missing features and when they are aprox planned. Everywhere is suggested to migrate and move to isolated worker. But this is not possible without this attribute. So currently the only option is to refactor all projects and dependencies to use .net 6 and remove all c# 11 features (required init) etc. and use in-proc. I do not want to sound rude or bring bad taste in here, i just need anything to at least plan with this blocker, that would be great! - Thanks a lot

**Comment by SeanFeldman** [comment]
@fabiocav, would it be possible to provide some insights into the plan for those requiring single execution? Thank you.

**Comment by BenjaBobs** [comment]
@essmd I wouldn't hold my breath.
I've managed to do a local work around [here](https://github.com/Azure/azure-functions-dotnet-worker/issues/938#issuecomment-1167329390), and it could potentially be expanded to work in a distributed manner if you write the mutex around blob/table storage instead of a `SemaphoreSlim`.

**Comment by essmd** [comment]
@BenjaBobs Thanks, i tested this and it works very well. Unfortunately i need distributed locking and also the support for the parameter binding. Highly appreciate your work as it helped me to get started! I could refactor your implementation a bit and working on those additional features. If i get something working, i will try to get a pull request ready for this repo. Maybe we can get this feature back a bit earlier this way :)

**Comment by wouter-b** [comment]
@fabiocav @jeffhollan Are there any plans to include something like the singleton attribute for the isolated model. At the moment this is the only issue holding us back from moving forward to NET8 and isolated functions.

**Comment by graham-saunders** [comment]
Any updates here? We heavily rely on singletons and its holding up a large portion of our .NET 8 migration.

**Comment by GonzaloVisma** [comment]
Is [Singleton] referencing using Microsoft.Azure.WebJobs in a Microsoft.Azure.Functions.Worker function working ? Just as workaround

**Comment by Basvg** [comment]
Just came across this roadmap update for .NET. The good news is that they are working on adding support for the in-process model with .NET 8. So once that is released, we can at least upgrade to .NET 8 and keep using the singletons for a few more years. Hopefully singleton support will be added to the isolated model at a later time.

https://techcommunity.microsoft.com/t5/apps-on-azure-blog/net-on-azure-functions-march-2024-roadmap-update/ba-p/4097744

**Comment by SeanFeldman** [comment]
This is a serious blocker for clients with ASB triggers when migrating off In-Process SDK.
[Workarounds](https://github.com/Azure/azure-functions-dotnet-worker/issues/938#issuecomment-1167329390) are not helpful as they worsen the situation.

**Comment by mattchenderson** [comment]
Hi all - the Singleton attribute also has mixed results in in-proc. Here's a host issue in which some of those challenges are captured: https://github.com/Azure/azure-functions-host/issues/912

In general, the singleton attribute as it exists in in-proc is just taking a distributed lock over some scope, applied to the triggering logic (specifically pre-dispatch), using blobs in AzureWebJobsStorage to do it. This means that other lifecycle considerations, such as service-specific state management or full handling of batches themselves are not presently incorporated. Those may not impact all scenarios, but they are a barrier to formalizing broader support for that even within in-proc; the feature is in an odd support state there, and we need to revisit that more holistically. But, for those that have not been impacted by those limitations, an answer is still needed. Right now, without a full feature, you can drive singleton behavior of your function processing logic - this is not the same as driving the trigger itself - using either:

- A function middleware that similarly takes a lock against storage for a given scope
- Stateful entities in Durable Functions to get similar lock processing

As noted both here and also in the thread Sean linked, these solutions would be _after_ the trigger has occurred, which is a key difference from the in-process model. We need an answer that governs the trigger itself. We're absolutely in agreement on that as a requirement.

As I write this, we're still discussing a few specifics to lock in our plan. Enabling the existing scope would be upleveling the support for the core singleton behavior that we know has problems. While a broader redesign around that would be nice, that's going to take more time and frankly isn't on the table as an answer to the current need. But we are still trying to figure out the right way to enable the existing behavior with an eye toward being able to fix that that at a later date. This is just a bit of detail work, and I'm hoping we can close that out soon and then knock this item out. We'll share the plan here once locked.

It would be helpful to get a few additional specifics around scenarios here, just so we can make sure they're properly accounted for in validation, etc. ServiceBusTrigger is clearly top of mind, and the pattern that works best in the in-process model would be with the function receiving a full batch or with the batch configured to have one message (so the function can just be given that one message).

**Comment by rory-a-hunt** [comment]
Hi @mattchenderson, I have a specific scenario which might help explain the need for singleton. I have a number of function apps which have fanned out to run micro batch processing on multiple shaker or instances, however I need to fan these back in and report status to a transactional source (SQL DB, snowflake etc) we don't want this happening on multiple instances, and the persistence source is also quite slow, so an efficient singular report/status process is required. 

Code has previous used SingletonMode.Listener on dotnet 6 inproc function apps. The upgrade to dotnet 8 is presenting this challenge. We are currently considering moving to a whole additional function app with constrained scale out and message handling, but that feels pretty compromised. 
Would be very glad to hear dev moving forward on this.

**Comment by bvmeer** [comment]
[@mattchenderson] We use the singleton attribute also for batch processing where we need to add some delay between the message handling. One of the receiving systems doesn't allow to much calls within a short period of time. So handling 1 by 1 guaranteed the amount off calls is within the limit each minute. 
We can't use scheduled messages for this. Because when one of the remote systems is down we need to wait till the other system is online again. In that case we pause the handling by disable the function and we pickup 1 by 1 after it switch online again.

**Comment by czb182** [comment]
This is a pretty big gap IMO.  My case is ServiceBusTrigger functions, to serialize execution, to avoid data concurrency issues.  The actual trigger supports own concurrency via maxConcurrent calls but that applies to the entire function app(via host.json), where in my case I want to apply to individual functions within an app.  Ideally it should also support scopeId like in process model.

**Comment by jimmytricks** [comment]
@mattchenderson, would also like to say that we make heavy use of singleton.

Our use case is that when Service A successfully sends, it triggers a retry of any failed messages via event grid. The retry pulls all messages from a blob and then puts them back into the service bus, so singleton is perfect as the recovery function needs to first read all the failed messages from a blob, put them into a service bus and delete them from the blob without other instances of this function running at the same time. Otherwise we have race conditions, duplicates and potentially a death loop without other workarounds.

We could fix this in other ways but singleton provides a really nice solution. Good to see there might be some progress on this, would you have an idea of a timeline? Cheers!

**Comment by keith-knight** [comment]
We also use the singleton attribute. The lack of support is currently blocking us from upgrading to NET 8. An ETA would be appreciated. 

I see that in-proc support for NET 8 has recently been added so it looks like it might not be far away?

https://github.com/Azure/azure-functions-host/issues/9951

**Comment by axiom-nz** [comment]
I have a slightly different use case.  In a multi tenanted application, we have the need for a single instance of an azure storage queue triggered function, per tenant. While its possible to perform something similar with Durable Functions, there is a lot that needs to be reworked in order to do so. 

Example
```
[Singleton(Mode = SingletonMode.Function, Scope = "tenant:{tenantId}")]
```

**Comment by jcageman** [comment]
> We also use the singleton attribute. The lack of support is currently blocking us from upgrading to NET 8. An ETA would be appreciated.
> 
> I see that in-proc support for NET 8 has recently been added so it looks like it might not be far away?
> 
> [Azure/azure-functions-host#9951](https://github.com/Azure/azure-functions-host/issues/9951)

same here! Please give an ETA.

**Comment by bn-spires** [comment]
i wanted to note that we are also blocked from moving to .NET 8 isolated, since we heavily rely on Singleton locking mechanism.  whats the ETA on that being supported?  Also does anyone have any valid proven out workarounds for this?

**Comment by jcageman** [comment]
> i wanted to note that we are also blocked from moving to .NET 8 isolated, since we heavily rely on Singleton locking mechanism. whats the ETA on that being supported? Also does anyone have any valid proven out workarounds for this?

Basically making your own distributed singleton lock with for example this https://github.com/fbeltrao/azfunctions-distributed-locking.

**Comment by SeanFeldman** [comment]
@jcageman [won't work](https://github.com/Azure/azure-functions-dotnet-worker/issues/274#issuecomment-2045835642) for message-based triggers.

**Comment by marklio** [comment]
I, too, have rate-limited resources that I manage with singleton. I'm using a blob lease to manually serialize, but it makes diagnostics really messy with so many triggers waiting and potentially timing out their queue message.

**Comment by sherman89** [comment]
Same issue here... Our solution is littered with singletons and my migration to isolated model PR is probably never going to be merged 😆 

EDIT: Are Service Bus sessions a good alternative to Singleton? We are only using singletons in SB triggered functions.

**Comment by sherman89** [comment]
In-process model will not be supported next year... Surely there has to be an official migration path??? 🤔

**Comment by bvmeer** [comment]
Due the lack of updates and we like to move forward to .Net10 now it is released.  We have a solution for our use-case, instead of a servicebus trigger we created a timer trigger, which is singleton by default. We let that timer trigger fetch for a new servicebus message. So we are sure that we only fetch 1 message at the time. And you can configure a timer to run each 30 seconds for example it would resolve our issue. 

> [[@mattchenderson](https://github.com/mattchenderson)] We use the singleton attribute also for batch processing where we need to add some delay between the message handling. One of the receiving systems doesn't allow to much calls within a short period of time. So handling 1 by 1 guaranteed the amount off calls is within the limit each minute. We can't use scheduled messages for this. Because when one of the remote systems is down we need to wait till the other system is online again. In that case we pause the handling by disable the function and we pickup 1 by 1 after it switch online again.

**Comment by jviau** [comment]
Chiming in here. Reading through the use cases my immediate response is "use durable". Not because it is a trigger available today (and we won't put in the effort to support singleton), but because I believe it is a better fit for the job than a singleton attribute. I understand having to rewrite to another trigger type is not ideal, but I believe you will be better served in the end by adopting durable.

You will be able to achieve singleton behavior through either entities or orchestrations. It will depend on your use case. Getting into implementation details here will be a bit much given the varied use cases described, but I am confident entities and orchestrations can accomplish singleton behavior better than any attribute would. Durable might even alleviate the need for a singleton-paradigm altogether and naturally better support your use case.

https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-overview?tabs=in-process%2Cnodejs-v3%2Cv1-model&pivots=csharp

**Comment by jviau** [comment]
Going to close this as not planned. Per my comment above, durable is better suited for these scenarios. Additionally, SingletonAttribute was never _intended_ to be supported in functions. It incidentally worked for in-proc due to it being present in WebJobs, but we have not added it to any out of proc worker yet.

If there are still concerns with migrations due to lack of Singleton, or questions about using durable, please open a new issue.
